ptf/saimpls.py

```diff
@@ -0,0 +1,3955 @@
+        send_packet(self, self.dev_port10, mpls_pkt)
+        verify_packet(self, recv_pkt, self.dev_port12)
+
+        sai_thrift_set_inseg_entry_attribute(
+            self.client, self.inseg_entry_1000,
+            pop_ttl_mode=SAI_INSEG_ENTRY_POP_TTL_MODE_UNIFORM)
+
+        recv_pkt['IPv6'].hlim = 54
+        print("Send MPLS tag packet after changing ttl mode to uniform")
+        send_packet(self, self.dev_port10, mpls_pkt)
+        verify_packet(self, recv_pkt, self.dev_port12)
+
+    def mplsEgressLERNullTermTtlModeTest(self):
+        '''
+        Verifies if MPLS null label is popped in Egress LER and packet
+        is forwarded based on IP lookup using the TTL value from the
+        popped MPLS label.
+        '''
+        print("\nmplsEgressLERNullTermTtlModeTest()")
+
+        send_pkt = simple_tcpv6_packet(
+            eth_dst=ROUTER_MAC,
+            ipv6_dst='200::1',
+            ipv6_hlim=64)
+        mpls_tag = {'label': 0, 'ttl': 60, 'tc':  0, 's':  1}
+        mpls_tag_list = [mpls_tag]
+        mpls_pkt = simple_mpls_packet(
+            eth_dst=ROUTER_MAC,
+            mpls_tags=mpls_tag_list,
+            inner_frame=send_pkt['IPv6'])
+        recv_pkt = simple_tcpv6_packet(
+            eth_dst='00:11:22:33:44:55',
+            eth_src=ROUTER_MAC,
+            ipv6_dst='200::1',
+            ipv6_hlim=59)
+
+        print("Send MPLS tag packet with label 0 - term and IP lookup")
+        send_packet(self, self.dev_port10, mpls_pkt)
+        verify_packet(self, recv_pkt, self.dev_port12)
+
+    def mplsEgressPhpTtlModeTest(self):
+        '''
+        Verifies PHP pops label and forwards packet using TTL
+        value from the top label in the stack.
+        '''
+        print("\nmplsEgressPhpTtlModeTest()")
+
+        send_pkt = simple_tcpv6_packet(
+            eth_dst=ROUTER_MAC,
+            ipv6_dst='300::1',
+            ipv6_hlim=64)
+
+        mpls_tag = {'label': 2000, 'ttl': 60, 'tc':  0, 's':  1}
+        mpls_tag_list = [mpls_tag]
+        mpls_pkt = simple_mpls_packet(
+            eth_dst=ROUTER_MAC,
+            mpls_tags=mpls_tag_list,
+            inner_frame=send_pkt['IPv6'])
+
+        mpls_tag_2 = {'label': 2111, 'ttl': 58, 'tc':  0, 's':  0}
+        mpls_tag_list_2 = [mpls_tag_2, mpls_tag]
+        mpls_tag_2_pkt = simple_mpls_packet(
+            eth_dst=ROUTER_MAC,
+            mpls_tags=mpls_tag_list_2,
+            inner_frame=send_pkt['IPv6'])
+
+        mpls_tag_3 = {'label': 2222, 'ttl': 56, 'tc':  0, 's':  0}
+        mpls_tag_list_3 = [mpls_tag_3, mpls_tag_2, mpls_tag]
+        mpls_tag_3_pkt = simple_mpls_packet(
+            eth_dst=ROUTER_MAC,
+            mpls_tags=mpls_tag_list_3,
+            inner_frame=send_pkt['IPv6'])
+
+        recv_pkt = simple_tcpv6_packet(
+            eth_dst='00:11:22:33:44:55',
+            eth_src=ROUTER_MAC,
+            ipv6_dst='300::1',
+            ipv6_hlim=63)
+        # Note: mpls_tag ttl is copied from top label, but it is not
```

Comment:
@mickeyspiegel could you tell me what you meant in this comment? Should the ttl in lines 3247 and 3255 be equal to 55 and 59?

Reply:
@pkochmanska

3247 should be 55, but that is not what we return. We should change or remove this part of the test.

In the p4 code we tied ttl decrement to a swap operation. At that point in the ttl code we no longer know whether any labels were popped earlier. If the test were to pop 3 labels and swap to explicit null (resulting in one new explicit null label remaining on the stack), then we would decrement to 55. I was suggesting to change the test to do that.

3255 is correct at 60 since that is pipe mode rather than uniform mode.

Note that this same issue also appears in the IPv4 tests.

There is another issue in mplsEgressPhpTtlModeTest at line 3286, where hlim should be 62 but we are returning 63. We should probably remove that part of the test.

Reply:
By changing just the value, you left the test in a state where we will fail the test. We need to remove these parts of the test.